### PR TITLE
Port linearize_gencost and pmin modifications from REISE.jl

### PR DESCRIPTION
**Pull request by jenhagg**
### Purpose
To facilitate using a grid.pkl and remove the need for mapping between a case.mat/grid.mat/etc, we move grid modifications previously done in REISE.jl to the scenario preparation process. 

This PR should be merged at the same time as https://github.com/Breakthrough-Energy/REISE.jl/pull/187, after which we should update the existing grid.mat files. I believe we also need to update the ramp30 values in plant.csv as discussed in this issue https://github.com/Breakthrough-Energy/PowerSimData/issues/680 and release a new dataset on zenodo.

### What the code is doing
Scale plant Pmin values and linearize cost curves the same way we were doing in REISE.jl - after we apply the change table, and just before we upload the grid.pkl to the server (or, container, or wherever). I also ported the linearize_gencost tests from julia since those will be unused/deleted.

### Testing
Ran a scenario in a container and reloaded it after it finished for some cursory spot checking.

### Time estimate
20 min

### CODE REVIEW

**Comment by rouille** [comment]
Is **powersimdata/design/generation/** a good location for the `cost_curves` module? Should we move it in **powersimdata/scenario/** or **powersimdata/input/**? What do you think @BainanXia?

File: powersimdata/scenario/execute.py [hunk 1]
```diff
@@ -203,9 +207,30 @@ def __init__(self, data_access, scenario_info, grid, ct):
...
+        plant.loc[plant.type.isin(profile_resource), "Pmin"] = 0
+        plant.loc[plant.status == 0, "Pmin"] = 0
+
     def prepare_grid(self):
         """Prepare grid for simulation."""
         print("--> Preparing grid data")
+        self._adjust_pmin()
+        storage = self.grid.storage
+        self.grid.gencost["after"] = linearize_gencost(self.grid)
+        storage["gencost"] = _linearize_gencost(storage["gencost"], storage["gen"])
```
Comment (rouille) [inline powersimdata/scenario/execute.py:232]: I am not sure I understand why we need two functions. Could we do:
```
storage["gencost"] = _linearize_gencost(storage["gencost"], storage["gen"])
self.grid.gencost["after"] = _linearize_gencost(elf.grid.gencost["before"], self.grid.plant)
```
Then we could discard the `linearize_gencost` function in the `cost_curves` module and rename `_linearize_gencost` to `linearize_gencost`

**Comment by BainanXia** [comment]
> Is **powersimdata/design/generation/** a good location for the `cost_curves` module? Should we move it in **powersimdata/scenario/** or **powersimdata/input/**? What do you think @BainanXia?

I agree. The original idea of `design` module is a place for functions that are optionally used in the scenario workflow. However, linearization is essential given the engine can't deal with quadratic cost curves (at least for now) and similarly for the ramp module in the future. Thus we can create a new module in `powersimdata/scenario/`. Not sure what is a good name for it, maybe `configure`?

File: powersimdata/scenario/execute.py [hunk 2]
```diff
@@ -203,9 +207,30 @@ def __init__(self, data_access, scenario_info, grid, ct):
...
+        def _scale(x):
+            factor = pmin_factor.get(x.type)
+            return x.Pmin if factor is None else factor * x.Pmax
+
+        plant.Pmin = plant.apply(_scale, axis=1)
+
+        # set Pmin to 0 for generators that are off or profile based
+        profile_resource = list(mi.plants["profile_resources"])
+        plant.loc[plant.type.isin(profile_resource), "Pmin"] = 0
+        plant.loc[plant.status == 0, "Pmin"] = 0
```
Comment (BainanXia) [inline powersimdata/scenario/execute.py:224]: Maybe out of the scope of this PR, two high level questions:
1. `hydro` is one of `profile_resources`, the `Pmin` of which will be set 0 at the end, however the `pmin_factor` of `hydro` is grid dependent .
2. We will need to build user interface to pass user specified `pmin_as_share_of_pmax` (either plant based or plant type based or plant type * zone based) in the future, right?

File: powersimdata/design/generation/tests/test_cost_curves.py [hunk 3]
```diff
@@ -141,6 +142,85 @@
...
+    actual = linearize_gencost(mock_grid_gc, num_segments=2)
+    assert_frame_equal(expected_two_segment, actual, check_dtype=False)
+
+
+def test_linearize_gencost_pmin_equal_pmax():
+    plant = mock_grid_gc.plant.copy()
+    plant.Pmin = plant.Pmax
+    grid = MockGrid({"plant": plant.reset_index().to_dict(), "gencost_before": mock_gc})
+    actual = linearize_gencost(grid, num_segments=3)
+    assert_frame_equal(expected_all_equal, actual, check_dtype=False)
```
Comment (BainanXia) [inline powersimdata/design/generation/tests/test_cost_curves.py:222]: Thanks! We thought about adding such tests a while ago as in the very first version of implementation, dividing by zero wasn't avoided.

File: powersimdata/scenario/execute.py (hunk 2 above)
Comment (BainanXia) [reply powersimdata/scenario/execute.py:224]: 1. True. Just pointing out such entries are modified twice in the current process.
2. Agreed.

**Comment by lanesmith** [comment]
> I agree. The original idea of `design` module is a place for functions that are optionally used in the scenario workflow. However, linearization is essential given the engine can't deal with quadratic cost curves (at least for now) and similarly for the ramp module in the future. Thus we can create a new module in `powersimdata/scenario/`. Not sure what is a good name for it, maybe `configure`?

Is the ramp module essential? Especially once (if?) `plant.csv` gets updated, I'm not sure we would need the ramp-rate correction anymore.

**Comment by BainanXia** [comment]
> > I agree. The original idea of `design` module is a place for functions that are optionally used in the scenario workflow. However, linearization is essential given the engine can't deal with quadratic cost curves (at least for now) and similarly for the ramp module in the future. Thus we can create a new module in `powersimdata/scenario/`. Not sure what is a good name for it, maybe `configure`?
> 
> Is the ramp module essential? Especially once (if?) `plant.csv` gets updated, I'm not sure we would need the ramp-rate correction anymore.

We will need that when scaling existing plants as well as adding new plants without specifying ramp rates.

File: powersimdata/scenario/execute.py (hunk 1 above)
Comment (jenhagg) [reply powersimdata/scenario/execute.py]: done

**Comment by rouille** [comment]
> > Is **powersimdata/design/generation/** a good location for the `cost_curves` module? Should we move it in **powersimdata/scenario/** or **powersimdata/input/**? What do you think @BainanXia?
> 
> I agree. The original idea of `design` module is a place for functions that are optionally used in the scenario workflow. However, linearization is essential given the engine can't deal with quadratic cost curves (at least for now) and similarly for the ramp module in the future. Thus we can create a new module in `powersimdata/scenario/`. Not sure what is a good name for it, maybe `configure`?

It can be done in a separate PR. I am fine with having the `linearize_gencost` function in a new `powersimdata.scenario.configure`. We could have the ramp rate setting there too. Or we could have a `linearize_cost_curve` module and a `assign_ramp_rate` modules in `powersimdata.input`. Hat is your opinion @jenhagg?

**Comment by jenhagg** [comment]
> > > Is **powersimdata/design/generation/** a good location for the `cost_curves` module? Should we move it in **powersimdata/scenario/** or **powersimdata/input/**? What do you think @BainanXia?
> > 
> > 
> > I agree. The original idea of `design` module is a place for functions that are optionally used in the scenario workflow. However, linearization is essential given the engine can't deal with quadratic cost curves (at least for now) and similarly for the ramp module in the future. Thus we can create a new module in `powersimdata/scenario/`. Not sure what is a good name for it, maybe `configure`?
> 
> It can be done in a separate PR. I am fine with having the `linearize_gencost` function in a new `powersimdata.scenario.configure`. We could have the ramp rate setting there too. Or we could have a `linearize_cost_curve` module and a `assign_ramp_rate` modules in `powersimdata.input`. Hat is your opinion @jenhagg?

Based on the current package structure, I think having those in `powersimdata.input` makes sense, although the distinction between input and design isn't super clear to me. I'm fine with either option; there is always the possibility of reorganizing in the future.

**Review by lanesmith — APPROVED** [review]
Looks good to me

**Review by rouille — APPROVED** [review]
Thank you for tackling this!

**Review by BainanXia — APPROVED** [review]
Nice 🚀
